tempesta_fw/http.c

```diff
@@ -2341,85 +2350,92 @@ tfw_http_req_mark_error(TfwHttpReq *req, int status)
 }
 
 /**
- * Functions define logging and response behaviour during detection of
- * malformed or malicious messages. Mark client connection in special
- * manner to delay its closing until transmission of error response
- * will be finished.
+ * Function defines logging and response behaviour during detection of
+ * malformed or malicious messages. If reply to client must be sent,
+ * mark client connection in special manner to delay its closing until
+ * transmission of error response will be finished; otherwise, close
+ * connection immediately.
  */
 static void
-tfw_http_cli_error_resp_and_log(bool reply, bool nolog, TfwHttpReq *req,
-				int status, const char *msg)
+tfw_http_error_resp_and_log(bool reply, bool nolog, TfwHttpReq *req,
+			    int status, const char *msg, bool close)
 {
 	if (!nolog)
 		TFW_WARN_ADDR(msg, &req->conn->peer->addr);
 
 	if (reply) {
-		TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
-		tfw_connection_unlink_msg(req->conn);
-		spin_lock(&cli_conn->seq_qlock);
-		list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
-		spin_unlock(&cli_conn->seq_qlock);
+		if (!close) {
```

Comment:
Hm, not sure about that. There are `BUG_ON(list_empty(&req->msg.seq_list));` checks in `tfw_http_resp_fwd()` which are called in  `tfw_http_req_mark_error()` function. I expect to see a crash if `reply == true && close == true`.

May be you need to set TFW_HTTP_B_CONN_CLOSE flag for a request if `close == true`? Tempesta closes connection to clients when forwarding a response to request withe the flag set. See `tfw_http_resp_init_ss_flags()`.

Reply:
Functions `tfw_client_drop()` and `tfw_client_block()` are intended to be called only in client connection context, and only during request processing - before request was attached to connection's `seq_queue`. I've mentioned this in comment:
https://github.com/tempesta-tech/tempesta/blob/02aa662140907ffa58fa1f52f899c18aa21bc72a/tempesta_fw/http.c#L2389.
Functions `tfw_client_drop_close()` and `tfw_client_block_close()` are intended to be called in all other cases - when request has already been attached to connection's `seq_queue` - in server connection context the request is found right at the beginning of the response processing:
https://github.com/tempesta-tech/tempesta/blob/02aa662140907ffa58fa1f52f899c18aa21bc72a/tempesta_fw/http.c#L3349
So  the request is in the `seq_queue` and `BUG_ON` in `tfw_http_resp_fwd()` should not be triggered. 
The same behavior was present before - but in `tfw_http_srv_error_resp_and_log` function, which I removed to optimize the code.

> May be you need to set TFW_HTTP_B_CONN_CLOSE flag for a request if close == true

This behavior is already implemented in `tfw_http_req_mark_error()` via `TFW_HTTP_F_SUSPECTED` flag.